clang, cc-wrapper: Move `--gcc-toolchain` logic into CC wrapper

Ericson2314 opened the pull request:
###### Motivation for this change

The new condition is slightly different, but I think it's OK?

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Ericson2314 committed:
clang, cc-wrapper: Move `--gcc-toolchain` logic into CC wrapper

Mic92 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -265,6 +265,7 @@ stdenv.mkDerivation {
     ##
     + optionalString (isClang && libcxx == null && !(stdenv.targetPlatform.useLLVM or false) && gccForLibs != null) ''
 
+      echo "--gcc-toolchain=${gccForLibs}" >> $out/nix-support/cc-cflags
```
It is intended to have this on macOS?

Ericson2314 replied:
On macOS `libcxx != null` so it shouldn't. On the other hand it might be needed with `libcxxClang` and not have it.

We might need to also make cc-wrapper libgcc/compiler-rt aware (something we probably should do in general) to fix that case if it is indeed messed up for `libcxxClang`.

Mic92 replied:
Ok. I was just asking if there is a libgcc based stdenv for macOS like their is a libcxx stdenv for Linux.

Ericson2314 replied:
Ah. I don't think there is. If there was it probably *would* use `--gcc-toolchain`. (When build finally build all the GCC libraries in separate derivations, this will be much easier!)

Mic92 replied:
Sounds good.

LnL7 replied:
There is a gccStdenv, which doesn't seem to use libc++ from what I can tell. I'm not sure what the implications of this flag are or if the condition triggers however.

```
HARDENING: disabled flags: pie
HARDENING: Is active (not completely disabled with "all" flag)
HARDENING: enabling fortify
HARDENING: enabling stackprotector
HARDENING: enabling strictoverflow
HARDENING: enabling format
HARDENING: enabling pic
extra flags before to /nix/store/q8ar8q8f2h4sq5563pcn9dnv7a8n8gns-gcc-9.3.0/bin/g++:
  -O2
  -D_FORTIFY_SOURCE=2
  -fstack-protector-strong
  --param
  ssp-buffer-size=4
  -fno-strict-overflow
  -Wformat
  -Wformat-security
  -Werror=format-security
  -fPIC
  -Wl\,-sdk_version
  -Wl\,10.12
  -Wl\,-no_uuid
original flags to /nix/store/q8ar8q8f2h4sq5563pcn9dnv7a8n8gns-gcc-9.3.0/bin/g++:
  -o
  hello
  hello.cc
extra flags after to /nix/store/q8ar8q8f2h4sq5563pcn9dnv7a8n8gns-gcc-9.3.0/bin/g++:
  -B/nix/store/fikpi6kfpfywadbqnrn6ldwif2glqvii-gcc-9.3.0-lib/lib
  -B/nix/store/i09q4si3r6ad3iblrdjxfmvl8bks6amg-Libsystem-osx-10.12.6/lib/
  -idirafter
  /nix/store/i09q4si3r6ad3iblrdjxfmvl8bks6amg-Libsystem-osx-10.12.6/include
  -idirafter
  /nix/store/q8ar8q8f2h4sq5563pcn9dnv7a8n8gns-gcc-9.3.0/lib/gcc/x86_64-apple-darwin/9.3.0/include-fixed
  -B/nix/store/nsq53ag0qb8q8ljxdwfxnm3k4ssbwqcf-gcc-wrapper-9.3.0/bin/
  -isystem
  /nix/store/29sdvdbaj16kxgzhxj72pa1dn3vrf9bw-clang-7.1.0/include
  -isystem
  /nix/store/29sdvdbaj16kxgzhxj72pa1dn3vrf9bw-clang-7.1.0/include
  -L/nix/store/29sdvdbaj16kxgzhxj72pa1dn3vrf9bw-clang-7.1.0/lib
  -L/nix/store/3z5szay7ax2p102ysnx8sgq8d305mp06-clang-7.1.0-lib/lib
  -L/nix/store/29sdvdbaj16kxgzhxj72pa1dn3vrf9bw-clang-7.1.0/lib
  -L/nix/store/3z5szay7ax2p102ysnx8sgq8d305mp06-clang-7.1.0-lib/lib
  -L/nix/store/i09q4si3r6ad3iblrdjxfmvl8bks6amg-Libsystem-osx-10.12.6/lib
  -L/nix/store/fikpi6kfpfywadbqnrn6ldwif2glqvii-gcc-9.3.0-lib/lib
```

Ericson2314 commented:
OK I rebased this, and added a `isLinux` condition to hew close to what we had before at the cost of some extra complexity, a long with a comment explaining the situation and how we might fix it better.

vcunat commented:
This PR was reverted (for now).  Unfortunately it isn't possible to reopen the original PR, but you can create a new one.
